commodity-utilities.scm -> c++

christopherlam opened the pull request:
step 1 of many for speed up currency conversions for bug 799258: make `gnc_get_match_commodity_splits` a cpp function. why does report eg asset chart fail?

```
* 00:18:52  WARN <gnc.gui> [gnc_run_report_with_error_handling()] Error in report: In ice-9/boot-9.scm:
  1752:10 11 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
          10 (apply-smob/0 #<thunk 7a3fdf514300>)
In c-interface.scm:
     34:5  9 (gnc:call-with-error-handling _ _)
In ice-9/boot-9.scm:
  1747:15  8 (with-exception-handler #<procedure 7a3fcb9258a0 at ic…> …)
  1752:10  7 (with-exception-handler _ _ #:unwind? _ # _)
In c-interface.scm:
    38:40  6 (_)
In report-core.scm:
   748:25  5 (gnc:report-render-html #<<report> type: "e9cf815f79db…> …)
In standard/category-barchart.scm:
   282:26  4 (category-barchart-renderer #<<report> type: "e9cf815f…> …)
In commodity-utilities.scm:
   824:31  3 (gnc:case-exchange-time-fn _ _ _ _ _ _)
   192:29  2 (gnc:get-commoditylist-totalavg-prices (#<swig-poin…> …) …)
In unknown file:
           1 (gnc-get-match-commodity-splits (#<swig-pointer Acc…> …) …)
In ice-9/boot-9.scm:
  1685:16  0 (raise-exception _ #:continuable? _)

In procedure gnc-get-match-commodity-splits: Wrong type argument in position 1: (#<swig-pointer Account * 17a7080> #<swig-pointer Account * 17a6e00> #<swig-pointer Account * 17a9b90> #<swig-pointer Account * 17a9d80> #<swig-pointer Account * 17a73d0> #<swig-pointer Account * 17abc90> #<swig-pointer Account * 17ac340> #<swig-pointer Account * 17a75c0> #<swig-pointer Account * 17aa240> #<swig-pointer Account * 17a77b0> #<swig-pointer Account * 17a7c40> #<swig-pointer Account * 17a8890> #<swig-pointer Account * 17a7eb0> #<swig-pointer Account * 17a8cf0> #<swig-pointer Account * 17ac7b0> #<swig-pointer Account * 17acde0> #<swig-pointer Account * 17a8f60> #<swig-pointer Account * 17a8280> #<swig-pointer Account * 17acfd0> #<swig-pointer Account * 17ad1c0> #<swig-pointer Account * 17ad3b0> #<swig-pointer Account * 17ad790> #<swig-pointer Account * 17ad5a0> #<swig-pointer Account * 17ad980> #<swig-pointer Account * 17adb70> #<swig-pointer Account * 17add60> #<swig-pointer Account * 17adf50>)
```

christopherlam committed:
[engine.i] gnc_get_match_commodity_splits from scheme to c++

The scheme version is very inefficient: (1) it uses QofQuery to
retrieve all splits, and QofQuery is slow. (2) The whole splitlist
typically is thousands splits long needs to be converted from C to
guile via consing and reverse. (3) The scheme list is immediately
filtered to select "interesting" splits only.

This function performs all of the above using Account's efficient
methods, selecting only interesting splits into a GList. This list is
then converted to scheme via cons and reverse.

christopherlam committed:
[commodity-utilities.scm] use gnc_get_match_commodity_splits

christopherlam commented:
Ok function converted to cpp is much more memory-friendly than guile, because in scheme the `get-all-splits` would retrieve _all_ 10'000s of splits, filtered into the 100s of foreign conversion splits; converted to c++ which does filtering within the lambda. This does seem to work.

christopherlam reviewed (commented):
don't sort unless requested

christopherlam committed:
Bug 799258 - Reports calculating net worth incorrectly after stock split

For average-cost and weighted-average: when the exchange-fn is
created, it would formerly consider the whole book foreign currency
conversions to calculate the date-specific conversions. This bug fix
will limit the book conversions to those earlier than the
date-specific conversions.

The book retrieval functions are expensive, therefore the analysis
splitlists are cached in the date-hash hash-table, which is scoped
within the exchange-fn lambda only, and is garbage collected when
exchange-fn is out of scope.

Also: amend relevant commodity-utilities.scm tests

The old weighted-average and average-cost tests were written to
document the (knowingly) incorrect exchanged amounts as originally
calculated pre-2024.

In 2024, the exchange-fn is fixed so that foreign currency
conversions at a particular report date would consider only
transactions up to the relevant report date.

The first AAPL transaction in this test book is on 09/08/2013: a
purchase of 600 AAPL at $36840 total @ $61.40 each.

Therefore the first test, on 20/02/2012 the AAPL->USD exchange rate
conversion should be zero, since there are no AAPL transactions
yet.

Subsequent AAPL test on 20/02/2014 shows the conversion of 1
AAPL to be 307/5 which matches $61.40.

etc.

christopherlam commented:
The new function is in Account.cpp because I can't fix `report.i` to swig `gnc_get_match_commodity_splits` properly. Anyway:

1. `gnc_get_match_commodity_splits` will scan all book splits and return only splits involving foreign exchange. Previously the whole book splitlist would be generated, scanned, filtered, gc'd.
2. `gnc:case-exchange-time-fn` which creates an `exchange-fn` can now accept a hash-table which will store the date-specific exchange-data. The hash-table should be scoped within a report render only.
3. The relevant multi-date reports amended to use the upgraded `gnc:case-exchange-time-fn`

Both 1 and 2 should mean multiyear reports would be much faster and would fix 799258 to boot...

jralls commented on gnucash/report/commodity-utilities.scm:
```diff
@@ -819,24 +820,26 @@
+(define* (gnc:case-exchange-time-fn
+          source-option report-currency commodity-list to-date-tp
+          start-percent delta-percent #:key date-hash)
+  (define-syntax-rule (memoize date expensive-fn)
```
What is `memoize` and where is it documented? I can't find it in the Guile docs.

christopherlam replied:
Its the definition right there. It is a macro.

christopherlam replied:
You could copy the whole `gnc:case-exchange-time-fn` definition, and paste into guile repl after `,expand`:

```scheme
cheme@(guile-user)> ,expand (define* (gnc:case-exchange-time-fn
          source-option report-currency commodity-list to-date-tp
          start-percent delta-percent #:key date-hash)
  (define-syntax-rule (memoize date expensive-fn)
    (if date-hash
        (or (hash-ref date-hash date) (hashv-set! date-hash date expensive-fn))
        expensive-fn))
  (case source-option
    ;; Make this the same as gnc:case-exchange-fn
    ((average-cost) (lambda (foreign domestic date)
                      ((memoize date (gnc:make-exchange-function
                                      (gnc:make-exchange-cost-alist
                                       report-currency date)))
                       foreign domestic)))
    ((weighted-average) (lambda (foreign domestic date)
                          (gnc:exchange-by-pricealist-nearest
                           (memoize date (gnc:get-commoditylist-totalavg-prices
                                          commodity-list report-currency date
                                          start-percent delta-percent))
                           foreign domestic date)))
    ((pricedb-before) gnc:exchange-by-pricedb-nearest-before)
    ((pricedb-latest) (lambda (foreign domestic date)
                        (gnc:exchange-by-pricedb-latest foreign domestic)))
    ((pricedb-nearest) gnc:exchange-by-pricedb-nearest)
    (else
     (begin
       (gnc:warn "gnc:case-exchange-time-fn: bad price-source value: "
                 source-option ". Using pricedb-nearest.")
       ;; FIX-ME another hack to prevent report crashing when an
       ;; unimplemented source-option comes through
       gnc:exchange-by-pricedb-nearest))))
$1 = (define* (gnc:case-exchange-time-fn
          source-option
          report-currency
          commodity-list
          to-date-tp
          start-percent
          delta-percent
          #:key
          (date-hash #f #:date-hash))
  (let ((key source-option))
    (cond (((@@ (guile) memv) key '(average-cost))
           (lambda (foreign domestic date)
             ((if date-hash
                (or (hash-ref date-hash date)
                    (hashv-set!
                      date-hash
                      date
                      (gnc:make-exchange-function
                        (gnc:make-exchange-cost-alist
                          report-currency
                          date))))
                (gnc:make-exchange-function
                  (gnc:make-exchange-cost-alist
                    report-currency
                    date)))
              foreign
              domestic)))
          (((@@ (guile) memv) key '(weighted-average))
           (lambda (foreign domestic date)
             (gnc:exchange-by-pricealist-nearest
               (if date-hash
                 (or (hash-ref date-hash date)
                     (hashv-set!
                       date-hash
                       date
                       (gnc:get-commoditylist-totalavg-prices
                         commodity-list
                         report-currency
                         date
                         start-percent
                         delta-percent)))
                 (gnc:get-commoditylist-totalavg-prices
                   commodity-list
                   report-currency
                   date
                   start-percent
                   delta-percent))
               foreign
               domestic
               date)))
          (((@@ (guile) memv) key '(pricedb-before))
           gnc:exchange-by-pricedb-nearest-before)
          (((@@ (guile) memv) key '(pricedb-latest))
           (lambda (foreign domestic date)
             (gnc:exchange-by-pricedb-latest foreign domestic)))
          (((@@ (guile) memv) key '(pricedb-nearest))
           gnc:exchange-by-pricedb-nearest)
          (else
           (gnc:warn
             "gnc:case-exchange-time-fn: bad price-source value: "
             source-option
             ". Using pricedb-nearest.")
           gnc:exchange-by-pricedb-nearest))))
```

jralls replied:
And what would that do for me?

christopherlam replied:
It illustrates why I inserted the macro. To avoid calling expensive-fn unless it's necessary.

jralls commented:
> I can't fix report.i to swig gnc_get_match_commodity_splits properly

Probably because you didn't change `swig-report` to write C++. `swig-engine` is already C++.

christopherlam commented:
> > I can't fix report.i to swig gnc_get_match_commodity_splits properly
> 
> Probably because you didn't change `swig-report` to write C++. `swig-engine` is already C++.

If you could move the function to gnc-report.cpp it would be a tremendous speedup and memory+gc improvement already.

christopherlam commented:
> If you could move the function to gnc-report.cpp it would be a tremendous speedup and memory+gc improvement already.

@jralls Do you think you could do this for me? i.e. move `gnc_get_match_commodity_splits` to `gnc-report.cpp`. The first 2 commits in this branch are good to go.

jralls commented:
> @jralls Do you think you could do this for me? i.e. move gnc_get_match_commodity_splits to gnc-report.cpp. The first 2 commits in this branch are good to go.

I have worked on it a bit yesterday and today and I'm leaning towards it really belonging in sw_engine rather than sw_reports because it exposes a bit too much of Account's internals and consequently needs to wrap Account*, AccountList*, and SplitList* that are already wrapped by sw_engine. Why do you think it would be better or faster in sw_report than in sw_engine?

christopherlam commented:
> > @jralls Do you think you could do this for me? i.e. move gnc_get_match_commodity_splits to gnc-report.cpp. The first 2 commits in this branch are good to go.
> 
> I have worked on it a bit yesterday and today and I'm leaning towards it really belonging in sw_engine rather than sw_reports because it exposes a bit too much of Account's internals and consequently needs to wrap Account*, AccountList*, and SplitList* that are already wrapped by sw_engine. Why do you think it would be better or faster in sw_report than in sw_engine?

Umm because engine is getting rather polluted, and this is a "analyze and don't modify splits" type function

christopherlam commented:
> > > @jralls Do you think you could do this for me? i.e. move gnc_get_match_commodity_splits to gnc-report.cpp. The first 2 commits in this branch are good to go.
> > 
> > 
> > I have worked on it a bit yesterday and today and I'm leaning towards it really belonging in sw_engine rather than sw_reports because it exposes a bit too much of Account's internals and consequently needs to wrap Account*, AccountList*, and SplitList* that are already wrapped by sw_engine. Why do you think it would be better or faster in sw_report than in sw_engine?
> 
> Umm because engine is getting rather polluted, and this is a "analyze and don't modify splits" type function

instead of swig-report it could go to swig-app-utils... but even then, i failed to swig it.

christopherlam committed:
[balsheet-pnl.scm] use enhanced gnc:case-exchange-time-fn

Because gnc:case-exchange-time-fn now handles multiple report dates
correctly.

jralls commented:
> Umm because engine is getting rather polluted, and this is a "analyze and don't modify splits" type function

I don't know what you mean by "polluted", but modification or lack thereof isn't significant. The function itself needs access to Engine internals, and swigging it requires guile&lt;-&gt;C/C++ conversions that live in bindings, in particular converting a Scheme list of SWIG_p_account_s into a GList of Account and a GList of Split into a Scheme list of SWIG_p_split_s. Combine that with a hefty serving of Spaghetti GnuCashese and you wind up having to duplicate a huge part of sw_engine in whatever other swig module you try to put it in on top of making the spaghetti problem worse by leaking engine implementation details.

jralls commented on gnucash/report/test/test-commodity-utils.scm:
```diff
@@ -655,15 +655,15 @@
                           #f #f)))
         (test-equal "gnc:case-exchange-time-fn weighted-average 20/02/2012"
-          307/5
+          0
```
The commit message needs to explain why the answers changed, otherwise it looks like you've done something wrong and patched the test to hide it.

christopherlam replied:
I'll amend when rewriting the commit message.

christopherlam replied:
test fix commit message amended.

jralls commented on gnucash/report/commodity-utilities.scm:
```diff
@@ -173,16 +173,16 @@
+            ;;        (gnc:make-gnc-monetary price-commodity share-amt))
+            ;;       ") =? "
+            ;;       (gnc:monetary->string
+            ;;        (gnc:make-gnc-monetary report-currency 0)))
```
Why are you removing these warnings? What has it got to do with caching the splits?

christopherlam replied:
These warnings will slow down the scanning of Forex splits. If the user chooses a multi date report eg net worth line chart spanning 10 years at daily resolution, the warnings will be shown 3650 times give or take.

jralls replied:
So set a flag and print the error once. Arguably it should abort the report because it will be wrong, but it might be friendlier to print a message on the report.

Regardless it should be its own commit.

christopherlam replied:
ok it'll be tricky because it'll be the _last_ report-date which captures the most splits which matters the most.

christopherlam replied:
My thoughts about showing the warnings once per report: we'd need to augment the case-exchange-date-fn to accept a "show-warnings?" optional argument which scans the splits and dumps any warnings. This would need to be called by reports once per report render.

christopherlam replied:
solved.

jralls reviewed (changes requested):
Otherwise looks mostly OK.

christopherlam committed:
[commodity-utilities.scm] routines upgraded to accept a #:hide-warnings? argument

... whose default is #f, therefore warnings are shown.

In gnc:case-exchange-time-fn the routines are called at the report
end-date once. Therefore all relevant splits are analysed for
warnings.

Subsequent calls (at various report dates) will omit warnings.

christopherlam commented:
so:

1. keep it in swig-engine but function lives in engine.i ?
2. add `gnc_account_foreach_split` which takes a std::function?
3. augment `gnc:case-exchange-time-fn` which accepts a `#:show-warnings?` optarg to show these warnings?

jralls reviewed (approved):
Much better. Thanks for separating the rework for easier review, don't forget to squash out the original versions before merging.

christopherlam committed:
[base-typemaps.i] add VECTOR_HELPER_INOUT for Split* and Account*

efficiently convert std::vector<Obj> to/from guile list of objects

christopherlam commented:
> Much better. Thanks for separating the rework for easier review, don't forget to squash out the original versions before merging.

Squashed up, also added `SplitVec` and `AccountVec` in base-typemaps.i which could shave off a few more nanoseconds and could be used going forward...
